dali/pipeline/executor/executor_factory.cc

```diff
@@ -58,40 +58,44 @@ bool ForceExec2() {
 }  // namespace
 
 template <typename... T>
-std::unique_ptr<ExecutorBase> GetExecutorImpl(bool pipelined, bool separated, bool async,
-                                              T&&... args) {
-  if (async && separated && pipelined) {
+std::unique_ptr<ExecutorBase> GetExecutorImpl(
+      bool pipelined, bool separated, bool async, bool dynamic, T&&... args) {
+  // Go over possible combinations and throw otherwise.
+  if (async && separated && pipelined && !dynamic) {
     return std::make_unique<AsyncSeparatedPipelinedExecutor>(std::forward<T>(args)...);
   } else if (async && !separated && pipelined) {
-    if (ForceExec2()) {
-      std::cerr << "\n!!! Forced use of Executor 2.0 !!!" << std::endl;
+    bool force_exec2 = ForceExec2();
+    if (dynamic || force_exec2) {
+      if (force_exec2)
+        std::cerr << "\n!!! Forced use of Executor 2.0 !!!" << std::endl;
       return std::make_unique<exec2::Executor2>(MakeExec2Config(std::forward<T>(args)...));
     } else {
      return std::make_unique<AsyncPipelinedExecutor>(std::forward<T>(args)...);
     }
-  } else if (!async && separated && pipelined) {
+  } else if (!async && separated && pipelined && !dynamic) {
     return std::make_unique<SeparatedPipelinedExecutor>(std::forward<T>(args)...);
-  } else if (!async && !separated && pipelined) {
+  } else if (!async && !separated && pipelined && !dynamic) {
     return std::make_unique<PipelinedExecutor>(std::forward<T>(args)...);
-  } else if (!async && !separated && !pipelined) {
+  } else if (!async && !separated && !pipelined && !dynamic) {
     return std::make_unique<SimpleExecutor>(std::forward<T>(args)...);
```

Comment:
Maybe all `!dynamic` could be moved to a separate branch:
```
if (!dynamic) {
  if (async && separated && pipelined) {
  } else if (!async && separated) {
  } else if (!async && !separated && pipelined) {
  } else if (!async && !separated && !pipelined) {
  }
  FAIL
}
else if (async && !separated && pipelined) {
}
else {
  FAIL
}```

Reply:
I don't know.... this way might be more verbose, but I think it's harder to get wrong.